### ISSUE TITLE
Deploying Multiple VMs with Managed Disks - Azure

### ISSUE THREAD

**Issue by sams976** [hashicorp/terraform-provider-azurerm#10905]
I followed a great blog by Neal Shah to create multiple disks on multiple VMs.
https://www.nealshah.dev/posts/2020/05/terraform-for-azure-deploying-multiple-vms-with-multiple-managed-disks/

Creating multiple VMs with multiple managed disks works well but I receive errors when trying to attach the managed disks to the VMs.

the line in question is 

```hcl
virtual_machine_id = azurerm_windows_virtual_machine.vm[element(split("_", each.key), 1)].id
```

When I use terraform console, I am able to successfully split and use the element.
```hcl
element(split("_", “datadisk_vm1_disk00”), 1)
“vm1”
```
I’m unsure if this is a bug or syntax issue

Thanks,

Sam

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.7
provider registry.terraform.io/hashicorp/azurerm v2.48.0

### Affected Resource(s)

* `azurerm_virtual_machine_data_disk_attachment`

### Terraform Configuration Files

variables.tf
```hcl

locals {
  vm_datadiskdisk_count_map = { for k in toset(var.vm_template1["vmname"]) : k => var.vm_template1["nb_disks_per_instance"] }
  luns                      = { for k in local.datadisk_lun_map : k.datadisk_name => k.lun }
  datadisk_lun_map = flatten([
    for vm_name, count in local.vm_datadiskdisk_count_map : [
      for i in range(count) : {
        datadisk_name = format("datadisk_%s_disk%02d", vm_name, i)
        lun           = i
      }
    ]
  ])
}

variable "azure_environment" {
  type = map(any)
  default = {
    location            = "australiasoutheast"
    resource_group_name = "terraformbuildrg"
  }
}

variable "vm_template1" {
  default = {
    vmname                = ["vm1", "vm2"]
    vmos                  = "windows"
    size                  = "Standard_D2_v4"
    storage_account_type  = "StandardSSD_LRS"
    os_disk_size          = "64"
    data_disk_size        = "32"
    nb_disks_per_instance = 1
  }
}

variable "images" {
  default = {
    windows = {
      publisher = "MicrosoftWindowsServer"
      offer     = "WindowsServer"
      sku       = "2016-Datacenter-smalldisk"
      version   = "latest"
    }
    linux = {
      publisher = "OpenLogic"
      offer     = "CentOS"
      sku       = "7.5"
      version   = "latest"
    }
    ubuntu = {
      publisher = "Canonical"
      offer     = "UbuntuServer"
      sku       = "18.04-LTS"
      version   = "latest"
    }
  }
}
```
createvm.tf
```hcl
resource "azurerm_network_interface" "nic" {
  count               = length(var.vm_template1["vmname"])
  name                = "${element(var.vm_template1["vmname"], count.index)}-nic"
  location            = var.azure_environment["location"]
  resource_group_name = var.azure_environment["resource_group_name"]

  ip_configuration {
    name = "ipconfig1"
    subnet_id = azurerm_virtual_network.vnet.subnet.*.id[1]
    private_ip_address_allocation = "Dynamic"
    # public_ip_address_id            = azurerm_public_ip.pubip.id
  }
}

# Create virtual machine Count
resource "azurerm_windows_virtual_machine" "vm" {
  count                 = length(var.vm_template1["vmname"])
  name                  = element(var.vm_template1["vmname"], count.index)
  location              = var.azure_environment["location"]
  resource_group_name   = var.azure_environment["resource_group_name"]
  size                  = var.vm_template1["size"]
  network_interface_ids = [element(azurerm_network_interface.nic.*.id, count.index)]
  admin_username = "adminuser"
  admin_password = "P@ssw0rd1234567!"
  tags = {
    environment = "Terraform"
  }

  os_disk {
    name                 = "${element(var.vm_template1["vmname"], count.index)}-osdisk"
    caching              = "ReadWrite"
    storage_account_type = var.vm_template1["storage_account_type"]
    disk_size_gb         = var.vm_template1["os_disk_size"]
  }

  source_image_reference {
    publisher = var.images[var.vm_template1["vmos"]]["publisher"]
    offer     = var.images[var.vm_template1["vmos"]]["offer"]
    sku       = var.images[var.vm_template1["vmos"]]["sku"]
    version   = var.images[var.vm_template1["vmos"]]["version"]
  }

  boot_diagnostics {
    storage_account_uri = azurerm_storage_account.diagsa.primary_blob_endpoint
  }

  # tags = var.tags
}

resource "azurerm_managed_disk" "managed_disk" {
   for_each             = toset([for j in local.datadisk_lun_map : j.datadisk_name])
   name                 = each.key
  location             = azurerm_resource_group.rg.location
  resource_group_name  = azurerm_resource_group.rg.name
  storage_account_type = "StandardSSD_LRS"
  create_option        = "Empty"
  disk_size_gb         = var.vm_template1["data_disk_size"]
  tags                 = var.vm_template1["tags"]
}

# Can't figure out why this isn't working
resource "azurerm_virtual_machine_data_disk_attachment" "managed_disk_attach" {
  for_each           = toset([for j in local.datadisk_lun_map : j.datadisk_name])
  managed_disk_id    = azurerm_managed_disk.managed_disk[each.key].id
  virtual_machine_id = azurerm_windows_virtual_machine.vm[element(split("_", each.key), 1)].id
  lun                = lookup(local.luns, each.key)
  caching            = "ReadWrite"
}
```

### Expected Behaviour

2 VMs each created with 1 managed disk

### Actual Behaviour

```hcl
Error: Invalid index

  on vmcountcreate.tf line 91, in resource "azurerm_virtual_machine_data_disk_attachment" "managed_disk_attach":
  91:   virtual_machine_id = azurerm_windows_virtual_machine.vm[element(split("_", each.key), 1)].id
    |----------------
    | azurerm_windows_virtual_machine.vm is tuple with 2 elements
    | each.key is "datadisk_vm1_disk00"

The given key does not identify an element in this collection value: a number
is required.


Error: Invalid index

  on vmcountcreate.tf line 91, in resource "azurerm_virtual_machine_data_disk_attachment" "managed_disk_attach":
  91:   virtual_machine_id = azurerm_windows_virtual_machine.vm[element(split("_", each.key), 1)].id
    |----------------
    | azurerm_windows_virtual_machine.vm is tuple with 2 elements
    | each.key is "datadisk_vm2_disk00"

The given key does not identify an element in this collection value: a number
is required.
```
### Steps to Reproduce

1. `terraform plan`

### References

* #0000

**Comment by sams976** [comment]
Is it possible to get the **virtual_machine_id** of a VM using the computer name? This could be my issue.

If not, how could I possible re-write the "azurerm_virtual_machine_data_disk_attachment" section so it will scale when I increare my VM from 2 to 3?

eg. vmname                = ["vm1", "vm2", "vm3"]

**Comment by ajithreddygeth** [comment]
@sams976  did you find any work around for this, I'm also facing the same issue

**Comment by sams976** [comment]
I went in a different direction when I needed to change the instance family for only vm1 but not vm2 or vm3.
In my varables.tf, I included a section for vm1, vm2 and vm3.
I then made createvm.tf portable so if I wanted a new vm4, I would clone createvm.tf to createvm4.tf
Opened createvm4.tf and performed a find & replace once to reference vm4 in variables.tf.

Its dirty but I'm still able to deploy a vm in under 2mins.

**Comment by sams976** [comment]
> @sams976 did you find any work around for this, I'm also facing the same issue

I went in a different direction when I needed to change the instance family for only vm1 but not vm2 or vm3.
In my varables.tf, I included a section for vm1, vm2 and vm3.
I then made createvm.tf portable so if I wanted a new vm4, I would clone createvm.tf to createvm4.tf
Opened createvm4.tf and performed a find & replace once to reference vm4 in variables.tf.

Its dirty but I'm still able to deploy a vm in under 2mins.
